# <center><h1>Companhia Aberta Demonstrativo Financeiro</h1></center>

# <h2>Load Libraries</h2>

In [493]:
import pandas as pd
import numpy as np

# <h2>Import Data</h2>

## <h3>Load .csv</h3>

In [494]:
cia_aberta_df = pd.read_csv(r'dfp_cia_aberta_BPA_con_2020.csv', encoding='ISO-8859-1', sep=";")    # leio o .csv com os dados de todas as companhias abertas

## <h3>Select Company</h3>

In [495]:
df = cia_aberta_df[cia_aberta_df['CNPJ_CIA'] == '97.837.181/0001-47'].copy()    # dataframe = linhas do dataframe onde ('CNPJ_CIA' == '97.837.181/0001-47') == True
                                                                                    # (seleciono somente as linhas relativas a uma companhia de interesse)

In [496]:
df = df[['ORDEM_EXERC','CD_CONTA','DS_CONTA','VL_CONTA']]

In [497]:
df = df[df['ORDEM_EXERC'] == 'ÚLTIMO']                                         # seleciono somente o último ou penúltimo exercício

In [498]:
df.reset_index(inplace=True, drop=True)

In [499]:
df

ORDEM_EXERC       CD_CONTA  \
0       ÚLTIMO              1   
1       ÚLTIMO           1.01   
2       ÚLTIMO        1.01.01   
3       ÚLTIMO        1.01.02   
4       ÚLTIMO     1.01.02.01   
..         ...            ...   
71      ÚLTIMO  1.02.04.02.07   
72      ÚLTIMO  1.02.04.02.08   
73      ÚLTIMO  1.02.04.02.09   
74      ÚLTIMO  1.02.04.02.10   
75      ÚLTIMO  1.02.04.02.11   

                                             DS_CONTA    VL_CONTA  
0                                         Ativo Total  11498520.0  
1                                    Ativo Circulante   4220022.0  
2                       Caixa e Equivalentes de Caixa   1728413.0  
3                              Aplicações Financeiras         0.0  
4   Aplicações Financeiras Avaliadas a Valor Justo...         0.0  
..                                                ...         ...  
71          Goodwill na aquisição da Caetex Florestal      8767.0  
72  Goodwill na aquisição da Cerâmica Urussanga em...     92944.0  
73  Goodwill na aquisição da Massima Revestimentos...      6110.0  
74  Goodwill na aquisição da Cecrisa Revestimentos...    168430.0  
75                               Goodwill Viva Decora         0.0  

[76 rows x 4 columns]

<h4>Subset DataFrame (only for testing purposes)</h4>

In [500]:
df = df.iloc[np.r_[0:7, 24:31],:]

In [501]:
df

ORDEM_EXERC       CD_CONTA  \
0       ÚLTIMO              1   
1       ÚLTIMO           1.01   
2       ÚLTIMO        1.01.01   
3       ÚLTIMO        1.01.02   
4       ÚLTIMO     1.01.02.01   
5       ÚLTIMO  1.01.02.01.01   
6       ÚLTIMO  1.01.02.01.02   
24      ÚLTIMO           1.02   
25      ÚLTIMO        1.02.01   
26      ÚLTIMO     1.02.01.01   
27      ÚLTIMO  1.02.01.01.01   
28      ÚLTIMO     1.02.01.02   
29      ÚLTIMO     1.02.01.03   
30      ÚLTIMO     1.02.01.04   

                                             DS_CONTA    VL_CONTA  
0                                         Ativo Total  11498520.0  
1                                    Ativo Circulante   4220022.0  
2                       Caixa e Equivalentes de Caixa   1728413.0  
3                              Aplicações Financeiras         0.0  
4   Aplicações Financeiras Avaliadas a Valor Justo...         0.0  
5                             Títulos para Negociação         0.0  
6                    Títulos Designados a Valor Justo         0.0  
24                               Ativo Não Circulante   7278498.0  
25                     Ativo Realizável a Longo Prazo   2071636.0  
26  Aplicações Financeiras Avaliadas a Valor Justo...         0.0  
27                   Títulos Designados a Valor Justo         0.0  
28  Aplicações Financeiras Avaliadas a Valor Justo...         0.0  
29  Aplicações Financeiras Avaliadas ao Custo Amor...         0.0  
30                                   Contas a Receber    124569.0

# <h2>Wrangling</h2>

## <h3>Number of Steps</h3>

In [502]:
cd_conta_split = pd.Series([string.split('.') for string in df['CD_CONTA']])

In [503]:
cd_conta_len = [len(lst) for lst in cd_conta_split]

In [504]:
num_levels = max(cd_conta_len)

In [505]:
num_levels

5

## <h3>ROUND 1</h3>

In [506]:
round=1

In [507]:
df

ORDEM_EXERC       CD_CONTA  \
0       ÚLTIMO              1   
1       ÚLTIMO           1.01   
2       ÚLTIMO        1.01.01   
3       ÚLTIMO        1.01.02   
4       ÚLTIMO     1.01.02.01   
5       ÚLTIMO  1.01.02.01.01   
6       ÚLTIMO  1.01.02.01.02   
24      ÚLTIMO           1.02   
25      ÚLTIMO        1.02.01   
26      ÚLTIMO     1.02.01.01   
27      ÚLTIMO  1.02.01.01.01   
28      ÚLTIMO     1.02.01.02   
29      ÚLTIMO     1.02.01.03   
30      ÚLTIMO     1.02.01.04   

                                             DS_CONTA    VL_CONTA  
0                                         Ativo Total  11498520.0  
1                                    Ativo Circulante   4220022.0  
2                       Caixa e Equivalentes de Caixa   1728413.0  
3                              Aplicações Financeiras         0.0  
4   Aplicações Financeiras Avaliadas a Valor Justo...         0.0  
5                             Títulos para Negociação         0.0  
6                    Títulos Designados a Valor Justo         0.0  
24                               Ativo Não Circulante   7278498.0  
25                     Ativo Realizável a Longo Prazo   2071636.0  
26  Aplicações Financeiras Avaliadas a Valor Justo...         0.0  
27                   Títulos Designados a Valor Justo         0.0  
28  Aplicações Financeiras Avaliadas a Valor Justo...         0.0  
29  Aplicações Financeiras Avaliadas ao Custo Amor...         0.0  
30                                   Contas a Receber    124569.0

### <h4>Join</h4>

In [508]:
cd_conta_split = pd.Series([string.split('.') for string in df['CD_CONTA']])

In [509]:
cd_conta_joincol = [row[:round] for row in cd_conta_split]

Select rows with only one element:

In [510]:
cd_conta_len = [len(row) for row in cd_conta_split]

In [511]:
idx_list = [idx for idx, element in enumerate(cd_conta_len) if element == round]        # round = 1

In [512]:
keys = [cd_conta_joincol[idx] for idx in idx_list]

In [513]:
key_idx_dict = {".".join(key): idx for key, idx in zip(keys, idx_list)}

In [514]:
df.iloc[idx_list]

ORDEM_EXERC CD_CONTA     DS_CONTA    VL_CONTA
0      ÚLTIMO        1  Ativo Total  11498520.0

Join:

In [515]:
ds_conta_1 = [df.DS_CONTA[key_idx_dict[key]]
              for key in key_idx_dict
              for row in cd_conta_joincol
              if ".".join(row) == key]

In [516]:
ds_conta_1 = [df.DS_CONTA[key_idx_dict[key]]
              for row in cd_conta_split
              for key in key_idx_dict
              if ".".join(row[:round]) == key]

In [517]:
df.insert(2, column='DS_CONTA_1', value=ds_conta_1)

In [518]:
df

ORDEM_EXERC       CD_CONTA   DS_CONTA_1  \
0       ÚLTIMO              1  Ativo Total   
1       ÚLTIMO           1.01  Ativo Total   
2       ÚLTIMO        1.01.01  Ativo Total   
3       ÚLTIMO        1.01.02  Ativo Total   
4       ÚLTIMO     1.01.02.01  Ativo Total   
5       ÚLTIMO  1.01.02.01.01  Ativo Total   
6       ÚLTIMO  1.01.02.01.02  Ativo Total   
24      ÚLTIMO           1.02  Ativo Total   
25      ÚLTIMO        1.02.01  Ativo Total   
26      ÚLTIMO     1.02.01.01  Ativo Total   
27      ÚLTIMO  1.02.01.01.01  Ativo Total   
28      ÚLTIMO     1.02.01.02  Ativo Total   
29      ÚLTIMO     1.02.01.03  Ativo Total   
30      ÚLTIMO     1.02.01.04  Ativo Total   

                                             DS_CONTA    VL_CONTA  
0                                         Ativo Total  11498520.0  
1                                    Ativo Circulante   4220022.0  
2                       Caixa e Equivalentes de Caixa   1728413.0  
3                              Aplicações Financeiras         0.0  
4   Aplicações Financeiras Avaliadas a Valor Justo...         0.0  
5                             Títulos para Negociação         0.0  
6                    Títulos Designados a Valor Justo         0.0  
24                               Ativo Não Circulante   7278498.0  
25                     Ativo Realizável a Longo Prazo   2071636.0  
26  Aplicações Financeiras Avaliadas a Valor Justo...         0.0  
27                   Títulos Designados a Valor Justo         0.0  
28  Aplicações Financeiras Avaliadas a Valor Justo...         0.0  
29  Aplicações Financeiras Avaliadas ao Custo Amor...         0.0  
30                                   Contas a Receber    124569.0

### <h4>Drop rows</h4>

Select rows with only one element:

In [519]:
df.iloc[idx_list]

ORDEM_EXERC CD_CONTA   DS_CONTA_1     DS_CONTA    VL_CONTA
0      ÚLTIMO        1  Ativo Total  Ativo Total  11498520.0

Check if there is at least another CD_CONTA starting with '1'

In [520]:
mask = [False if len == round else True for len in cd_conta_len]

In [521]:
any(row[0] == '1' for row in cd_conta_split[mask])

True

In [522]:
# cd_conta_split = cd_conta_split[mask]

In [523]:
df = df[mask]

In [524]:
df.reset_index(inplace=True, drop=True)

In [525]:
df

ORDEM_EXERC       CD_CONTA   DS_CONTA_1  \
0       ÚLTIMO           1.01  Ativo Total   
1       ÚLTIMO        1.01.01  Ativo Total   
2       ÚLTIMO        1.01.02  Ativo Total   
3       ÚLTIMO     1.01.02.01  Ativo Total   
4       ÚLTIMO  1.01.02.01.01  Ativo Total   
5       ÚLTIMO  1.01.02.01.02  Ativo Total   
6       ÚLTIMO           1.02  Ativo Total   
7       ÚLTIMO        1.02.01  Ativo Total   
8       ÚLTIMO     1.02.01.01  Ativo Total   
9       ÚLTIMO  1.02.01.01.01  Ativo Total   
10      ÚLTIMO     1.02.01.02  Ativo Total   
11      ÚLTIMO     1.02.01.03  Ativo Total   
12      ÚLTIMO     1.02.01.04  Ativo Total   

                                             DS_CONTA   VL_CONTA  
0                                    Ativo Circulante  4220022.0  
1                       Caixa e Equivalentes de Caixa  1728413.0  
2                              Aplicações Financeiras        0.0  
3   Aplicações Financeiras Avaliadas a Valor Justo...        0.0  
4                             Títulos para Negociação        0.0  
5                    Títulos Designados a Valor Justo        0.0  
6                                Ativo Não Circulante  7278498.0  
7                      Ativo Realizável a Longo Prazo  2071636.0  
8   Aplicações Financeiras Avaliadas a Valor Justo...        0.0  
9                    Títulos Designados a Valor Justo        0.0  
10  Aplicações Financeiras Avaliadas a Valor Justo...        0.0  
11  Aplicações Financeiras Avaliadas ao Custo Amor...        0.0  
12                                   Contas a Receber   124569.0

## <h3>ROUND 2</h3>

In [526]:
round=2

In [527]:
df

ORDEM_EXERC       CD_CONTA   DS_CONTA_1  \
0       ÚLTIMO           1.01  Ativo Total   
1       ÚLTIMO        1.01.01  Ativo Total   
2       ÚLTIMO        1.01.02  Ativo Total   
3       ÚLTIMO     1.01.02.01  Ativo Total   
4       ÚLTIMO  1.01.02.01.01  Ativo Total   
5       ÚLTIMO  1.01.02.01.02  Ativo Total   
6       ÚLTIMO           1.02  Ativo Total   
7       ÚLTIMO        1.02.01  Ativo Total   
8       ÚLTIMO     1.02.01.01  Ativo Total   
9       ÚLTIMO  1.02.01.01.01  Ativo Total   
10      ÚLTIMO     1.02.01.02  Ativo Total   
11      ÚLTIMO     1.02.01.03  Ativo Total   
12      ÚLTIMO     1.02.01.04  Ativo Total   

                                             DS_CONTA   VL_CONTA  
0                                    Ativo Circulante  4220022.0  
1                       Caixa e Equivalentes de Caixa  1728413.0  
2                              Aplicações Financeiras        0.0  
3   Aplicações Financeiras Avaliadas a Valor Justo...        0.0  
4                             Títulos para Negociação        0.0  
5                    Títulos Designados a Valor Justo        0.0  
6                                Ativo Não Circulante  7278498.0  
7                      Ativo Realizável a Longo Prazo  2071636.0  
8   Aplicações Financeiras Avaliadas a Valor Justo...        0.0  
9                    Títulos Designados a Valor Justo        0.0  
10  Aplicações Financeiras Avaliadas a Valor Justo...        0.0  
11  Aplicações Financeiras Avaliadas ao Custo Amor...        0.0  
12                                   Contas a Receber   124569.0

### <h4>Join</h4>

In [528]:
cd_conta_split = pd.Series([string.split('.') for string in df['CD_CONTA']])

In [529]:
cd_conta_joincol = [row[:round] for row in cd_conta_split]

Select rows with only 2 elements:

In [530]:
cd_conta_len = [len(lst) for lst in cd_conta_split]

In [531]:
idx_list = [idx for idx, element in enumerate(cd_conta_len) if element == round]

In [532]:
df.iloc[idx_keys]

ORDEM_EXERC CD_CONTA   DS_CONTA_1              DS_CONTA   VL_CONTA
0      ÚLTIMO     1.01  Ativo Total      Ativo Circulante  4220022.0
6      ÚLTIMO     1.02  Ativo Total  Ativo Não Circulante  7278498.0

In [533]:
keys = [cd_conta_joincol[idx] for idx in idx_keys]

In [534]:
key_idx_dict = {".".join(key): idx for key, idx in zip(keys, idx_keys)}

In [535]:
ds_conta_2 = [df.DS_CONTA[key_idx_dict[key]]
              for key in key_idx_dict
              for row in cd_conta_joincol
              if ".".join(row) == key]

In [536]:
ds_conta_2 = [df.DS_CONTA[key_idx_dict[key]]
              for row in cd_conta_split
              for key in key_idx_dict
              if ".".join(row[:round]) == key]

In [537]:
# ds_conta_2 = [df['DS_CONTA'][key_idx_dict[key]] for row in cd_conta_split for key in key_idx_dict if ".".join(row[:round]) == key]

In [538]:
df.insert(3, column='DS_CONTA_2', value=ds_conta_2)

In [539]:
df

ORDEM_EXERC       CD_CONTA   DS_CONTA_1            DS_CONTA_2  \
0       ÚLTIMO           1.01  Ativo Total      Ativo Circulante   
1       ÚLTIMO        1.01.01  Ativo Total      Ativo Circulante   
2       ÚLTIMO        1.01.02  Ativo Total      Ativo Circulante   
3       ÚLTIMO     1.01.02.01  Ativo Total      Ativo Circulante   
4       ÚLTIMO  1.01.02.01.01  Ativo Total      Ativo Circulante   
5       ÚLTIMO  1.01.02.01.02  Ativo Total      Ativo Circulante   
6       ÚLTIMO           1.02  Ativo Total  Ativo Não Circulante   
7       ÚLTIMO        1.02.01  Ativo Total  Ativo Não Circulante   
8       ÚLTIMO     1.02.01.01  Ativo Total  Ativo Não Circulante   
9       ÚLTIMO  1.02.01.01.01  Ativo Total  Ativo Não Circulante   
10      ÚLTIMO     1.02.01.02  Ativo Total  Ativo Não Circulante   
11      ÚLTIMO     1.02.01.03  Ativo Total  Ativo Não Circulante   
12      ÚLTIMO     1.02.01.04  Ativo Total  Ativo Não Circulante   

                                             DS_CONTA   VL_CONTA  
0                                    Ativo Circulante  4220022.0  
1                       Caixa e Equivalentes de Caixa  1728413.0  
2                              Aplicações Financeiras        0.0  
3   Aplicações Financeiras Avaliadas a Valor Justo...        0.0  
4                             Títulos para Negociação        0.0  
5                    Títulos Designados a Valor Justo        0.0  
6                                Ativo Não Circulante  7278498.0  
7                      Ativo Realizável a Longo Prazo  2071636.0  
8   Aplicações Financeiras Avaliadas a Valor Justo...        0.0  
9                    Títulos Designados a Valor Justo        0.0  
10  Aplicações Financeiras Avaliadas a Valor Justo...        0.0  
11  Aplicações Financeiras Avaliadas ao Custo Amor...        0.0  
12                                   Contas a Receber   124569.0

### <h4>Drop rows</h4>

Select rows with only 2 elements:

In [540]:
df.iloc[idx_list]

ORDEM_EXERC CD_CONTA   DS_CONTA_1            DS_CONTA_2  \
0      ÚLTIMO     1.01  Ativo Total      Ativo Circulante   
6      ÚLTIMO     1.02  Ativo Total  Ativo Não Circulante   

               DS_CONTA   VL_CONTA  
0      Ativo Circulante  4220022.0  
6  Ativo Não Circulante  7278498.0

Check if there is at least another CD_CONTA starting with '1.01' or '1.02':

In [541]:
'''
for idx in idx_list:
    '''

'\nfor idx in idx_list:\n    '

In [542]:
idx_list[0]

0

In [543]:
mask = [True]*len(ds_conta_2)
mask[idx_list[0]] = False

In [544]:
any(row[:round] == cd_conta_joincol[idx_list[0]] for row in cd_conta_split[mask])

True

In [545]:
df = df.drop(idx_list[0])

In [546]:
idx_list[1]

6

In [547]:
mask = [True]*len(ds_conta_2)
mask[idx_list[1]] = False

In [548]:
any(row[:round] == cd_conta_joincol[idx_list[1]] for row in cd_conta_split[mask])

True

In [549]:
df = df.drop(idx_list[1])

In [550]:
df.reset_index(inplace=True, drop=True)

In [551]:
df

ORDEM_EXERC       CD_CONTA   DS_CONTA_1            DS_CONTA_2  \
0       ÚLTIMO        1.01.01  Ativo Total      Ativo Circulante   
1       ÚLTIMO        1.01.02  Ativo Total      Ativo Circulante   
2       ÚLTIMO     1.01.02.01  Ativo Total      Ativo Circulante   
3       ÚLTIMO  1.01.02.01.01  Ativo Total      Ativo Circulante   
4       ÚLTIMO  1.01.02.01.02  Ativo Total      Ativo Circulante   
5       ÚLTIMO        1.02.01  Ativo Total  Ativo Não Circulante   
6       ÚLTIMO     1.02.01.01  Ativo Total  Ativo Não Circulante   
7       ÚLTIMO  1.02.01.01.01  Ativo Total  Ativo Não Circulante   
8       ÚLTIMO     1.02.01.02  Ativo Total  Ativo Não Circulante   
9       ÚLTIMO     1.02.01.03  Ativo Total  Ativo Não Circulante   
10      ÚLTIMO     1.02.01.04  Ativo Total  Ativo Não Circulante   

                                             DS_CONTA   VL_CONTA  
0                       Caixa e Equivalentes de Caixa  1728413.0  
1                              Aplicações Financeiras        0.0  
2   Aplicações Financeiras Avaliadas a Valor Justo...        0.0  
3                             Títulos para Negociação        0.0  
4                    Títulos Designados a Valor Justo        0.0  
5                      Ativo Realizável a Longo Prazo  2071636.0  
6   Aplicações Financeiras Avaliadas a Valor Justo...        0.0  
7                    Títulos Designados a Valor Justo        0.0  
8   Aplicações Financeiras Avaliadas a Valor Justo...        0.0  
9   Aplicações Financeiras Avaliadas ao Custo Amor...        0.0  
10                                   Contas a Receber   124569.0

## <h3>ROUND 3</h3>

In [586]:
round=3

In [553]:
df

ORDEM_EXERC       CD_CONTA   DS_CONTA_1            DS_CONTA_2  \
0       ÚLTIMO        1.01.01  Ativo Total      Ativo Circulante   
1       ÚLTIMO        1.01.02  Ativo Total      Ativo Circulante   
2       ÚLTIMO     1.01.02.01  Ativo Total      Ativo Circulante   
3       ÚLTIMO  1.01.02.01.01  Ativo Total      Ativo Circulante   
4       ÚLTIMO  1.01.02.01.02  Ativo Total      Ativo Circulante   
5       ÚLTIMO        1.02.01  Ativo Total  Ativo Não Circulante   
6       ÚLTIMO     1.02.01.01  Ativo Total  Ativo Não Circulante   
7       ÚLTIMO  1.02.01.01.01  Ativo Total  Ativo Não Circulante   
8       ÚLTIMO     1.02.01.02  Ativo Total  Ativo Não Circulante   
9       ÚLTIMO     1.02.01.03  Ativo Total  Ativo Não Circulante   
10      ÚLTIMO     1.02.01.04  Ativo Total  Ativo Não Circulante   

                                             DS_CONTA   VL_CONTA  
0                       Caixa e Equivalentes de Caixa  1728413.0  
1                              Aplicações Financeiras        0.0  
2   Aplicações Financeiras Avaliadas a Valor Justo...        0.0  
3                             Títulos para Negociação        0.0  
4                    Títulos Designados a Valor Justo        0.0  
5                      Ativo Realizável a Longo Prazo  2071636.0  
6   Aplicações Financeiras Avaliadas a Valor Justo...        0.0  
7                    Títulos Designados a Valor Justo        0.0  
8   Aplicações Financeiras Avaliadas a Valor Justo...        0.0  
9   Aplicações Financeiras Avaliadas ao Custo Amor...        0.0  
10                                   Contas a Receber   124569.0

### <h4>Join</h4>

In [554]:
cd_conta_split = pd.Series([string.split('.') for string in df['CD_CONTA']])

In [593]:
cd_conta_joincol = [row[:round] for row in cd_conta_split]

Select rows with only 3 elements:

In [556]:
cd_conta_len = [len(lst) for lst in cd_conta_split]

In [557]:
idx_list = [idx for idx, element in enumerate(cd_conta_len) if element == round]

In [558]:
df.iloc[idx_list]

ORDEM_EXERC CD_CONTA   DS_CONTA_1            DS_CONTA_2  \
0      ÚLTIMO  1.01.01  Ativo Total      Ativo Circulante   
1      ÚLTIMO  1.01.02  Ativo Total      Ativo Circulante   
5      ÚLTIMO  1.02.01  Ativo Total  Ativo Não Circulante   

                         DS_CONTA   VL_CONTA  
0   Caixa e Equivalentes de Caixa  1728413.0  
1          Aplicações Financeiras        0.0  
5  Ativo Realizável a Longo Prazo  2071636.0

In [594]:
keys = [cd_conta_joincol[idx] for idx in idx_list]

In [560]:
key_idx_dict = {".".join(key): idx for key, idx in zip(keys, idx_list)}

In [561]:
ds_conta_3 = [df.DS_CONTA[key_idx_dict[key]]
              for row in cd_conta_split
              for key in key_idx_dict
              if ".".join(row[:round]) == key]

In [562]:
df.insert(4, column='DS_CONTA_3', value=ds_conta_3)

In [563]:
df

ORDEM_EXERC       CD_CONTA   DS_CONTA_1            DS_CONTA_2  \
0       ÚLTIMO        1.01.01  Ativo Total      Ativo Circulante   
1       ÚLTIMO        1.01.02  Ativo Total      Ativo Circulante   
2       ÚLTIMO     1.01.02.01  Ativo Total      Ativo Circulante   
3       ÚLTIMO  1.01.02.01.01  Ativo Total      Ativo Circulante   
4       ÚLTIMO  1.01.02.01.02  Ativo Total      Ativo Circulante   
5       ÚLTIMO        1.02.01  Ativo Total  Ativo Não Circulante   
6       ÚLTIMO     1.02.01.01  Ativo Total  Ativo Não Circulante   
7       ÚLTIMO  1.02.01.01.01  Ativo Total  Ativo Não Circulante   
8       ÚLTIMO     1.02.01.02  Ativo Total  Ativo Não Circulante   
9       ÚLTIMO     1.02.01.03  Ativo Total  Ativo Não Circulante   
10      ÚLTIMO     1.02.01.04  Ativo Total  Ativo Não Circulante   

                        DS_CONTA_3  \
0    Caixa e Equivalentes de Caixa   
1           Aplicações Financeiras   
2           Aplicações Financeiras   
3           Aplicações Financeiras   
4           Aplicações Financeiras   
5   Ativo Realizável a Longo Prazo   
6   Ativo Realizável a Longo Prazo   
7   Ativo Realizável a Longo Prazo   
8   Ativo Realizável a Longo Prazo   
9   Ativo Realizável a Longo Prazo   
10  Ativo Realizável a Longo Prazo   

                                             DS_CONTA   VL_CONTA  
0                       Caixa e Equivalentes de Caixa  1728413.0  
1                              Aplicações Financeiras        0.0  
2   Aplicações Financeiras Avaliadas a Valor Justo...        0.0  
3                             Títulos para Negociação        0.0  
4                    Títulos Designados a Valor Justo        0.0  
5                      Ativo Realizável a Longo Prazo  2071636.0  
6   Aplicações Financeiras Avaliadas a Valor Justo...        0.0  
7                    Títulos Designados a Valor Justo        0.0  
8   Aplicações Financeiras Avaliadas a Valor Justo...        0.0  
9   Aplicações Financeiras Avaliadas ao Custo Amor...        0.0  
10                                   Contas a Receber   124569.0

### <h4>Drop rows</h4>

Select rows with only 3 elements:

In [564]:
df.iloc[idx_list]

ORDEM_EXERC CD_CONTA   DS_CONTA_1            DS_CONTA_2  \
0      ÚLTIMO  1.01.01  Ativo Total      Ativo Circulante   
1      ÚLTIMO  1.01.02  Ativo Total      Ativo Circulante   
5      ÚLTIMO  1.02.01  Ativo Total  Ativo Não Circulante   

                       DS_CONTA_3                        DS_CONTA   VL_CONTA  
0   Caixa e Equivalentes de Caixa   Caixa e Equivalentes de Caixa  1728413.0  
1          Aplicações Financeiras          Aplicações Financeiras        0.0  
5  Ativo Realizável a Longo Prazo  Ativo Realizável a Longo Prazo  2071636.0

Check if there is at least another row starting with '1.01.01', '1.01.02' or '1.02.01':

In [ ]:
'''
for idx in idx_list:
    '''

In [587]:
idx_list[0]

0

In [588]:
mask = [True]*len(ds_conta_3)
mask[idx_list[0]] = False

In [589]:
any(row[:round] == cd_conta_joincol[idx_list[0]] for row in cd_conta_split[mask])

False

In [595]:
idx_list[1]

1

In [596]:
mask = [True]*len(ds_conta_3)
mask[idx_list[1]] = False

In [598]:
any(row[:round] == cd_conta_joincol[idx_list[1]] for row in cd_conta_split[mask])

True

In [599]:
df = df.drop(idx_list[1])

In [600]:
idx_list[2]

5

In [601]:
mask = [True]*len(ds_conta_3)
mask[idx_list[2]] = False

In [602]:
any(row[:round] == cd_conta_joincol[idx_list[1]] for row in cd_conta_split[mask])

True

In [603]:
df = df.drop(idx_list[2])

In [604]:
df.reset_index(inplace=True, drop=True)

In [605]:
df

ORDEM_EXERC       CD_CONTA   DS_CONTA_1            DS_CONTA_2  \
0      ÚLTIMO        1.01.01  Ativo Total      Ativo Circulante   
1      ÚLTIMO     1.01.02.01  Ativo Total      Ativo Circulante   
2      ÚLTIMO  1.01.02.01.01  Ativo Total      Ativo Circulante   
3      ÚLTIMO  1.01.02.01.02  Ativo Total      Ativo Circulante   
4      ÚLTIMO     1.02.01.01  Ativo Total  Ativo Não Circulante   
5      ÚLTIMO  1.02.01.01.01  Ativo Total  Ativo Não Circulante   
6      ÚLTIMO     1.02.01.02  Ativo Total  Ativo Não Circulante   
7      ÚLTIMO     1.02.01.03  Ativo Total  Ativo Não Circulante   
8      ÚLTIMO     1.02.01.04  Ativo Total  Ativo Não Circulante   

                       DS_CONTA_3  \
0   Caixa e Equivalentes de Caixa   
1          Aplicações Financeiras   
2          Aplicações Financeiras   
3          Aplicações Financeiras   
4  Ativo Realizável a Longo Prazo   
5  Ativo Realizável a Longo Prazo   
6  Ativo Realizável a Longo Prazo   
7  Ativo Realizável a Longo Prazo   
8  Ativo Realizável a Longo Prazo   

                                            DS_CONTA   VL_CONTA  
0                      Caixa e Equivalentes de Caixa  1728413.0  
1  Aplicações Financeiras Avaliadas a Valor Justo...        0.0  
2                            Títulos para Negociação        0.0  
3                   Títulos Designados a Valor Justo        0.0  
4  Aplicações Financeiras Avaliadas a Valor Justo...        0.0  
5                   Títulos Designados a Valor Justo        0.0  
6  Aplicações Financeiras Avaliadas a Valor Justo...        0.0  
7  Aplicações Financeiras Avaliadas ao Custo Amor...        0.0  
8                                   Contas a Receber   124569.0